### ISSUE TITLE
Support for LLVM 15?

### ISSUE THREAD

**Issue by woodruffw** [cdisselkoen/llvm-ir#24]
First of all, thank you for maintaining this crate -- it's been incredibly useful on some of our projects!

Recent versions of `rustc` have upgraded to LLVM 15, meaning that the IR generated by `rustc` can't be used with this crate's bindings by default.

Would it be possible to add LLVM 15 support to this crate? I'm happy to take a stab at it first as well, if it's easier for you to review changes 🙂

**Comment by cdisselkoen** [comment]
Yes, I will work on LLVM 15 support (or if you want to take the first stab, feel free).  LLVM 15 has some [major changes](https://releases.llvm.org/15.0.0/docs/ReleaseNotes.html#changes-to-the-llvm-ir) which will affect us -- in particular the move to opaque pointers, but also the removal of a number of constant expressions.

**Comment by woodruffw** [comment]
I'm happy to take a first stab! I can at least do the boilerplate bits, and I'm somewhat familiar with the IR-level opaque pointer changes.

**Comment by woodruffw** [comment]
Opened #25 as a start.

**Comment by cdisselkoen** [comment]
More discussion at #25.  I've added a bunch to @woodruffw's start, and there is now an `llvm-15` branch in this repo which represents the latest progress as far as I'm aware.  Unfortunately there is a segfault when running our test suite on the `llvm-15` branch, which I haven't had time to track down yet.  (My time for this project is limited, apologies.)

**Comment by another-rex** [comment]
Spent some time looking into this, the segfault seems to be caused by this line: 

https://github.com/cdisselkoen/llvm-ir/blob/94bb32cb4f7263d0e95d97d1d2146f947af08fe3/src/function.rs#LL327C42-L327C42

because in llvm15 now uses opaque pointers by default: https://releases.llvm.org/15.0.0/docs/ReleaseNotes.html#changes-to-the-llvm-ir

Following the [migration doc](https://releases.llvm.org/15.0.0/docs/ReleaseNotes.html#changes-to-the-llvm-ir) I replaced that line with:
```rust
let functy = unsafe { LLVMGlobalGetValueType(func) };
```

which seems to work. 

Most of the tests still need to be migrated, and still need to replace the `LLVMGetElementType` usages in `types.rs`.

**Comment by woodruffw** [comment]
Thanks @cdisselkoen, and everyone else as well! I'm excited to use these changes on our tools 🙂
